CI

toots opened the pull request.

whitequark commented on tests/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+RUN wget https://raw.github.com/ocaml/opam/master/shell/opam_installer.sh -O - | sh -s /usr/local/bin 4.04.0+32bit
+RUN opam repository add windows git://github.com/whitequark/opam-cross-windows
+RUN opam pin --no-action add ocamlbuild https://github.com/ocaml/ocamlbuild.git
```
I think we can use an ocamlbuild release already.

toots commented:
It's getting there and already showing interesting results. However, the `mxe` dependencies handling is too limited, I need to be able to go down the deps. @whitequark do you know how to use `mxe` dependencies in `opam`?

whitequark commented:
Does this help? https://github.com/whitequark/opam-cross-windows#external-dependencies

toots commented:
Absolutely!

toots commented:
Ok, I think this is ready to merge.

There are two scripts:
* `run_tests.sh` to build all packages, no argument
* `run_test.sh <package>` to build just one package
Default output is condensed. If you set `VERBOSE=true` in the environment, it should show some logs from the build.

Builds are done inside a docker image with `mxe` and cross-building tools for windows 32 bit. Should be fairly easy to add windows 64 bit.

Here's the current output of the build. Some packages fail for obvious reason (`ocaml-windows64`), at least fails because `libpng` isn't available for download when installing via `mxe`. At least one fails because it doesn't have a `mxe` dependency. The chunk of the failures comes from the `opam install` phase. Not sure why.

```
Building base image [ok]✅
Building astring-windows.0.8.3 [failed]🚫
Building atdgen-windows.1.10.0 [ok]✅
Building atd-windows.1.2.0 [ok]✅
Building base64-windows.2.1.2 [failed]🚫
Building biniou-windows.1.0.12 [ok]✅
Building bos-windows.0.1.4 [failed]🚫
Building calendar-windows.2.03.2 [ok]✅
Building camlbz2-windows.0.6.0 [ok]✅
Building camlgpc-windows.1.0 [ok]✅
Building camlimages-windows.4.2.2 [failed]🚫
Building camlzip-windows.1.07 [ok]✅
Building cmdliner-windows.0.9.8 [failed]🚫
Building conf-flambda-windows.1 [ok]✅
Building conf-gcc-windows.1 [ok]✅
Building conf-gcc-windows32.1 [ok]✅
Building conf-gcc-windows64.1 [failed]🚫
Building cryptokit-windows.1.10 [ok]✅
Building ctypes-windows.0.11.0 [ok]✅
Building dtools-windows.0.3.4 [ok]✅
Building easy-format-windows.1.2.0 [ok]✅
Building expat-windows.0.9.1 [ok]✅
Building expat-windows.0.9.1+dev [ok]✅
Building faad-windows.0.3.3 [ok]✅
Building fdkaac-windows.0.2.1 [ok]✅
Building flac-windows.0.1.3 [ok]✅
Building flexdll-windows.0.34 [ok]✅
Building fmt-windows.0.8.0 [failed]🚫
Building fpath-windows.0.7.1 [failed]🚫
Building gdal-windows.0.9.0 [failed]🚫
Building jsonm-windows.0.9.1 [failed]🚫
Building lablgl-windows.1.05 [ok]✅
Building lame-windows.0.3.3 [ok]✅
Building logs-windows.0.6.2 [failed]🚫
Building lwt_react-windows.1.0.1 [failed]🚫
Building lwt-windows.2.6.0 [failed]🚫
Building lwt-windows.3.0.0 [failed]🚫
Building lwt-zmq-windows.2.0.1 [failed]🚫
Building mad-windows.0.4.5 [ok]✅
Building magic-mime-windows.1.0.0 [ok]✅
Building mtime-windows.0.8.3 [failed]🚫
Building ocaml-windows32.4.04.0 [ok]✅
Building ocaml-windows.4.04.0 [ok]✅
Building ocaml-windows64.4.04.0 [failed]🚫
Building ocplib-endian-windows.1.0 [ok]✅
Building ogg-windows.0.5.2 [ok]✅
Building opus-windows.0.1.2 [ok]✅
Building proj4-windows.0.9.1 [ok]✅
Building react-windows.1.2.0 [failed]🚫
Building result-windows.1.2 [failed]🚫
Building re-windows.1.7.1 [ok]✅
Building rresult-windows.0.5.0 [failed]🚫
Building samplerate-windows.0.1.4 [ok]✅
Building speex-windows.0.2.1 [ok]✅
Building stdint-windows.0.3.0-0 [ok]✅
Building tgls-windows.0.8.4 [failed]🚫
Building theora-windows.0.3.1 [ok]✅
Building uint-windows.1.2.0 [ok]✅
Building uutf-windows.0.9.4 [failed]🚫
Building vorbis-windows.0.7.0 [ok]✅
Building yojson-windows.1.3.3 [ok]✅
Building zmq-windows.4.0-7 [failed]🚫
```

whitequark commented on README.md:
```diff
@@ -104,7 +104,7 @@ External dependencies
 opam-windows-cross is designed to use native dependencies from the [MXE environment](http://mxe.cc). It is possible to automatically install all required dependencies for an OPAM package, e.g. `camlbz2-windows`, using one short command within an MXE checkout:
 
-    make `opam list --short --recursive --external=mxe camlbz2-windows`
+    make `opam list --short --recursive --external=mxe --required-by=camlbz2-windows`
```
I thought `--recursive` should have been taking care of that?

toots replied:
From the opam doc:
```
       --recursive
           With `--depends-on' and `--required-by', display all transitive
           dependencies rather than just direct dependencies.
```

whitequark commented on packages/re-windows.1.7.1/opam:
```diff
@@ -6,4 +6,8 @@ build: [
   ["env" "OCAMLFIND_TOOLCHAIN=windows" "ocaml" "setup.ml" "-install"]
 ]
 remove: [["ocamlfind" "-toolchain" "windows" "remove" "re"]]
-depends: ["ocaml-windows" "ocamlfind"]
```
Can be shortened to `depends: ["ocaml-windows" "ocamlbuild" {build}]` since `ocaml-windows` implies `ocamlfind`.

toots replied:
Word.

whitequark commented:
Oh and can you please special-case ocaml-windows*?

toots commented:
@whitequark I've added a first try for circle CI but I think you need to enable it on the repository.

toots commented:
And, yeah, I plan on adding a `skip` section, with 64 bit stuff but also packages that won't build like `zmq` (no `mxe` dep)

whitequark commented:
Travis enabled now.

whitequark commented:
Okay, here are my wishes for the CI (aka what I think makes it usable):
  * Docker image caching (this seems to take a *lot* of time)
  * Two Travis CI build matrix entries; one builds only what's changed in the PR to see if we broke that, another builds everything to see whether we broke, well, everything.
  * Whichever fails the first brings down the build.

I've configured the master branch as protected. Meaning, Travis must return success for pull-request rebased on top of HEAD before anything is merged, and after one approved review everything is merged on its own.

Seems OK?

toots commented:
Seems good to me. My only concern is that we keep those tools as guidelines for decision making without making it too difficult in the cases when they need to be circumvented, which always happen at some point.

whitequark commented:
> Seems good to me. My only concern is that we keep those tools as guidelines for decision making without making it too difficult in the cases when they need to be circumvented, which always happen at some point.

Well, broken packages should get fixed or dropped from the index. What to do for packages that we can't verify is a more complex question. I vote for skipping them in the scripts until an obvious solution emerges.

toots commented:
For sure. We can see about it when it happens. I've had a couple of occasions lately with the `opam-packages` repository where the build would fail due to a bug in the specific version of OCaml used in the CI. There was no point taking on such a huge task -- updating the OCaml compiler version -- just to merge by packaging PR so it was merged even though the CI didn't green light it.

whitequark commented:
The cross-compilation repositories do not provide more than one compiler version though (the combinatorial explosion is a nightmare) so that can't happen here.

toots commented:
@whitequark, for caching, we'd need a dockerhub account to push to. I can do it with my own account but perhaps you want to create a opam-cross account?

toots commented:
Also, if you want to have a reasonable build time, I would advise against the "build the world" part. Even if we cache the base image with `ocaml-windows` installed, there's no way that build will be under 20 min I think. It just takes a while to recompile all these packages one by one.

Matter of fact, `opam` CI doesn't to the whole world build either, just a difference build..

toots commented:
Ok, I've done the following:
* Enabled caching via `ocamlcross/windows-x86-base`
* Added diff-based builds, including `revdeps`.
* Added optional `WORLD` and `BUILD_BASE` to, respectively, build all the packages and build the base image.

I recommend merging as it is. I want to move forward with #33. I already expect that this PR will fail since some of the revdeps there do not currently build so we'll have to decide what to do.

whitequark commented:
Thanks a lot for your work! I'll take a bit of time to review this, but won't ask for any more features.

Can you elaborate on the dockerhub account? I'd like to have access to it.

toots commented:
I created a dockerhub organization here: https://hub.docker.com/u/ocamlcross/dashboard/ if you give me your user there, I can add you to it.

whitequark commented:
`whitequark` as usual

toots commented:
Added!

whitequark commented:
Thanks for your work!

toots commented:
Awesome, thanks for merging. Now on to `jbuild` with cross-compilation support!
